### DISCUSSION TITLE
Deployment Strategies

### DISCUSSION THREAD

**Discussion by mikesmeltzer** [oqtane/oqtane.framework#887] [Q&A]
Hello,

Does anyone have any recommended strategies for deploying the site config / content between various environments? like Dev > QA > Etc.. 

For example, I can deploy my custom modules through a standard CI/CD process but what about the placements of the modules and the screens/pages, etc.? Since that is all dynamic in the database, migrations can be cumbersome. Would looking into Site Templates or selectively picking data from the DB with custom scripts, etc. be an appropriate way?

Any have experience / thoughts on this?

Thanks,

Mike

**Comment by mikecasas** [comment]
I've tried SQL scripts and the Site template approach. I would say the template approach is slightly better.

You can set the default site template in the appsettings file, so you can deploy different setups.

You raise a good point on migrations though...

Not sure I answered your question, but the more discussion on this the better?

How do you plan to deploy your site? To a VM, using Azure App Service or a container environment?

**Comment by mikesmeltzer** [comment]
Not a specific answer but more info on what other people is doing the better 👍 

I deploy most things to an Azure App Service and plan to with Oqtane too.

I'll look at the Site Templates and see if they might meet my needs or give me some ideas.

**Comment by sbwalker** [comment]
My suggestion would be to use SiteTemplates - however currently they are only utilized when you initially create your site. The goal was that in the longer term you could also utilize SiteTemplates for more granular migrations.

**Reply by sbwalker** [reply]
@mikesmeltzer what workflow do you use in DNN to perform migration between different environments?

**Reply by mikesmeltzer** [reply]
@sbwalker this has always been a pain point in DNN for me and my customers, how to move the content/config. In some instances, it's a manual effort, in some we use a combination of custom SQL scripts that target specific tables and in others we built a "staging" extension that would run sql scripts behind the scenes to copy the data to a target environment. 

My original question was due to this operational pain I've experienced in DNN and was hoping to see what others do in Oqtane or if Oqtane has something to help make that process a bit smoother.

**Reply by sbwalker** [reply]
It does not have a solution for this problem yet... and I don't think the SiteTemplate concept is the right approach for this use case. I can imagine a feature where you specify a Url to your lower environment, provide some configuration, and it retrieves the configuration/content and applies it to the higher environment.

**Reply by mikesmeltzer** [reply]
I agree, the Site Template doesn't fit this. The approach you mention sounds like a custom staging tool I built (and also the old DNN Enterprise Content Staging). 

Do you think this is something that the core framework should be handling or leave it up to extension developers or custom dev workflows to implement something? If it's a core framework item, would you envision this taking care of only the core content/config or also be extensible to 3rd party devs for custom extensions?

**Reply by sbwalker** [reply]
I am not sure at this point if this should be a core framework responsibility or not. The need for a modular application framework to allow for migrations/deployments from environment to environment is important. Whether or not this needs to include support for content is an open question as Oqtane is not a CMS. Certainly there will be the opportunity for a third party to build their own solution as well.

**Comment by markdav-is** [comment]
@sbwalker I watched a nice demo you ran for "page migrations" in version 3.1.  I'm wondering if you have some examples of that.  We have a dev environment where we build pages from scratch, but we have lots of pages that are very common.  I'd like to script those pages when we go to other sites. Site 1 = Dev on localhost, Site 2 = url in the cloud.  I was thinking about a SQL script but I'd prefer page migrations.

**Reply by sbwalker** [reply]
Did you look at the repo with the source code related to the 3.1 video? It contains an example of how to use Site Migrations to add a page to a specific site. 

https://github.com/oqtane/Oqtane.Demo31

**Reply by markdav-is** [reply]
awesome thanks!  "link in the description" of course, duh.
